### PR TITLE
Fix show code checkbox mismatch

### LINKED ISSUE #239: "Show code on hover" always checked on switch to authoring view, but not always working
1. Switch to authoring view, see "show code on hover" checked
2. Un-check "show code on hover"
3. Switch to notebook view, then back to authoring view
4. "Show code on hover" is checked again, but hover doesn't show the code
5. Un-check and then re-check the button and observe hover code is now visible

Either the code should be shown on hover in step 4 or the button should retain its state (if unchecked) when switching to authoring view.

### PR DESCRIPTION
**Pull request by parente**
* Always reset to true when switching modes
* Don't leak event listeners every time help HTML is added

Fixes #239

### CODE REVIEW

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js [hunk 1]
```diff
@@ -112,6 +120,8 @@ define([
                     if ($helpArea) {
                         // remove if it exists since layout-specific help text will be inserted
                         $helpArea.remove();
```
Comment (dalogsdon) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js]: Since we are using jQuery to remove the help template elements, are the changes to the event handler creation above necessary? It would be nice to keep them in the help template.

From https://api.jquery.com/remove/:
> Similar to .empty(), the .remove() method takes elements out of the DOM. Use .remove() when you want to remove the element itself, as well as everything inside it. In addition to the elements themselves, all bound events and jQuery data associated with the elements are removed. To remove the elements without removing data and events, use .detach() instead.

Comment (parente) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js]: I think `remove` only cleans up event handlers on the target element, the help area in this case. Any handlers registered on descendant elements (the checkbox and more info link in this case) leak I believe.

>The only way that a $.cache entry gets deleted is when you call remove() on the element that put it there!` 

from http://makandracards.com/makandra/31325-how-to-create-memory-leaks-in-jquery

I believe this is why jQuery support for bubbling exists. You register handlers on stable elements so you don't have to worry about registering and unregistering the listeners as content comes and goes.

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js [hunk 2]
```diff
@@ -69,6 +69,14 @@ define([
...
 
     PolymerSupport.init();
+    
+    // register delegate handlers that persist across dashboard mode changes  
+    $('#notebook_panel').on('change', '.show-code-label input', function() {
+        $('#notebook-container').toggleClass('hide-code', !this.checked);
+    });
+    $('#notebook_panel').on('click', '.more-info-link', function() {
+        $(this).parents('.help-area').toggleClass('show-more', 200);
+    });
```
Comment (jhpedemonte) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js]: What about this:

1. Keep event handlers in template. Locality of code is good and means we won't have to search through code if we change structure of template in the future.
2. Use event namespaces: i.e. `.on('change.DashboardHelpArea')`
3. Before calling `@helpArea.remove()`, call `.off('.DashboardHelpArea')` to remove all event handlers

Comment (parente) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js]: > 3. Before calling $helpArea.remove(), call .off('.DashboardHelpArea') to remove all event handlers

But you have to know exactly what node to call off on, which means you still need to know the structure of the template right?

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js (hunk 1 above)
Comment (parente) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js]: remove() appears to leaks descendent handlers.

https://jsfiddle.net/obL75u7k/4/

screen shot 2016-05-11 at 2 11 59 pm

Yellow is number of registered event handlers. Up, up up!

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js (hunk 2 above)
Comment (dalogsdon) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js]: Yes, but at least you have to know less.

Comment (parente) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js]: Well, no, you need to know exactly the same (which elements have the handlers to unregister), the code is two places (register in template vs unregister in JS), and it has to register and re-register every time the template is replaced (vs once in the JS).

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js (hunk 1 above)
Comment (dalogsdon) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js]: I ran that fiddle and got this result:

image

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js (hunk 2 above)
Comment (dalogsdon) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js]: I was thinking if there is a jQuery function to remove all namespaced events on all child elements. If that doesn't exist, then the elements do need to be known.

**Comment by dalogsdon** [comment]
Closing in favor of #246
